HTTP/3: Improve Kestrel doc (#22987)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/servers/kestrel/http3.md

# Use HTTP/3 with the ASP.NET Core Kestrel web server

> [HTTP/3](https://quicwg.org/base-drafts/draft-ietf-quic-http.html) is the third and upcoming major version of HTTP. HTTP/3 uses the same semantics as HTTP/1.1 and HTTP/2: the same request methods, status codes, and message fields apply to all versions. The differences are in the underlying transport. Both HTTP/1.1 and HTTP/2 use TCP as their transport. HTTP/3 uses a new transport technology developed alongside HTTP/3 called QUIC.

Reviewer: ```suggestion
[HTTP/3](https://quicwg.org/base-drafts/draft-ietf-quic-http.html) is the third and upcoming major version of HTTP. HTTP/3 uses the same semantics as HTTP/1.1 and HTTP/2: the same request methods, status codes, and message fields apply to all versions. The differences are in the underlying transport. Both HTTP/1.1 and HTTP/2 use TCP as their transport. HTTP/3 uses a new transport technology developed alongside HTTP/3 called [QUIC](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-34).
```

> HTTP/3 and QUIC have a number of benefits compared to older HTTP versions:

Reviewer: ```suggestion
HTTP/3 and QUIC have a number of benefits compared to HTTP/1.1 and HTTP/2:
```

* Faster response time of the first request. QUIC and HTTP/3 negotiates the connection in fewer round-trips between the client and the server. The first request reachs the server faster.

> * Improved experience when there is connection packet loss. HTTP/2 multiplexes multiple requests via one TCP connection. Packet loss on the connection would affect all requests. This problem is called "head-of-line blocking". Because QUIC provides native multiplexing, lost packets only impact the requests where data has been lost.

Reviewer: ```suggestion
* Improved experience when there is connection packet loss. HTTP/2 multiplexes multiple requests via one TCP connection. Packet loss on the connection affects all requests. This problem is called "head-of-line blocking". Because QUIC provides native multiplexing, lost packets only impact the requests where data has been lost.
```

> * Supports transitioning between networks. This feature is useful for mobile devices where it is common to switch between WIFI and cellular networks as a mobile device changes location. Today HTTP/1.1 and HTTP/2 connections will fail with an error and force an app or web browser to retry. HTTP/3 allows the app or web browser to seamlessly continue when a network changes. Kestrel doesn't support network transitions in .NET 6. It may be available in a future release.

Reviewer: ```suggestion
* Supports transitioning between networks. This feature is useful for mobile devices where it is common to switch between WIFI and cellular networks as a mobile device changes location. Currently HTTP/1.1 and HTTP/2 connections fail with an error and force an app or web browser to retry when switching from WIFI to cellular. HTTP/3 allows the app or web browser to seamlessly continue when a network changes. Kestrel doesn't support network transitions in .NET 6. It may be available in a future release.
```

Possibly

We hope to make it available in a future release.

> [!IMPORTANT]
> HTTP/3 is available in .NET 6 as a preview feature. The HTTP/3 specification hasn't been finalized. Additionally, there may be behavioral or performance issues in HTTP/3 with .NET 6.
>

> > Apps that want to use HTTP/3 should be designed to support it alongside older HTTP versions. If issues are identified in HTTP/3 then it is recommended that apps temporarily disable it until issues are resolved and the app can be updated.

Reviewer: ```suggestion
> Apps that want take advantage of HTTP/3 should be designed to support HTTP/3 and HTTP/2. If issues are identified in HTTP/3 then it is recommended that apps temporarily disable it until issues are resolved and the app can be updated.
```

or 

```suggestion
> Apps that want take advantage of HTTP/3 should be designed to support all 3 versions versions. If issues are identified in HTTP/3 then it is recommended that apps temporarily disable it until issues are resolved and the app can be updated.
```

or 

```suggestion
> Apps that want take advantage of HTTP/3 should be designed to support HTTP/3, HTTP/2, and possibly HTTP/1. If issues are identified in HTTP/3 then it is recommended that apps temporarily disable it until issues are resolved and the app can be updated.
```

## HTTP/3 requirements

> HTTP/3 isn't supported everywhere. The requirements are different depending on the operating system that Kestrel is running on.

Reviewer: ```suggestion
HTTP/3 isn't supported everywhere. The requirements are different depending on the operating system hosting Kestrel.
```

> When using HTTP/3, it is recommended to enable it alongside other HTTP versions. Because Kestrel only enables HTTP/3 if an environment supports it, a port configured to serve all HTTP protocols will automatically enable HTTP/3 if it is available. If HTTP/3 isn't available then the port continues to serve HTTP/1.1 and HTTP/2 requests.

Reviewer: ```suggestion
When using HTTP/3, we recommend enabling it and the other HTTP versions. Because Kestrel only enables HTTP/3 if an environment supports it, a port configured to serve all HTTP protocols automatically enable HTTP/3 if it is available. If HTTP/3 isn't available then the port continues to serve HTTP/1.1 and HTTP/2 requests.
```

> When using HTTP/3, we recommend enabling it and the other HTTP versions. Because Kestrel only enables HTTP/3 if an environment supports it, a port configured to serve all HTTP protocols automatically enable HTTP/3 if it is available. If HTTP/3 isn't available then the port continues to serve HTTP/1.1 and HTTP/2 requests.

Reviewer 2: This parts a little confusing. Kestrel doesn't automatically enable HTTP/3, you have to turn it on. It will no-op on platforms where it's not supported if a fallback is available.

Reviewer 2: It's also important to have a fallback protocol because an application is likely to experience connectivity issues for QUIC / HTTP/3 until all routers, firewalls, and proxies are updated to support it.

> When using HTTP/3, we recommend enabling it and the other HTTP versions. Because Kestrel only enables HTTP/3 if an environment supports it, a port configured to serve all HTTP protocols automatically enable HTTP/3 if it is available. If HTTP/3 isn't available then the port continues to serve HTTP/1.1 and HTTP/2 requests.

Author: ```suggestion
HTTP/3 isn't configured by default. Because not all routers, firewalls, and proxies properly support HTTP/3, we recommend configuring HTTP/3 together with the other HTTP versions. This can be done by specifying `HttpProtocols.Http1AndHttp2AndHttp3` as an endpoint's supported protocols.

If the platform that Kestrel is running on doesn't have all the requirements for HTTP/3 then it is automatically disabled.
```

Reviewer 2: "HTTP/3 isn't enabled by default."

But otherwise, 👍

> For example, `HttpProtocols.Http1AndHttp2AndHttp3` allows Kestrel to enable HTTP/3 on environments where it is supported, along side HTTP/1.1 and HTTP/2.

Reviewer: ```suggestion
For example, `HttpProtocols.Http1AndHttp2AndHttp3` allows Kestrel to enable HTTP/3 on environments where it is supported,  with fallbacks for HTTP/1.1 and HTTP/2.
```

* Supports transitioning between networks. This feature is useful for mobile devices where it is common to switch between WIFI and cellular networks as a mobile device changes location. Currently HTTP/1.1 and HTTP/2 connections fail with an error when switching networks. An app or web browsers must retry any failed HTTP requests. HTTP/3 allows the app or web browser to seamlessly continue when a network changes. Kestrel doesn't support network transitions in .NET 6. It may be available in a future release.

> [!IMPORTANT]
> HTTP/3 is available in .NET 6 as a preview feature. The HTTP/3 specification hasn't been finalized. Additionally, there may be behavioral or performance issues in HTTP/3 with .NET 6.
>
> Apps that want take advantage of HTTP/3 should be designed to support all 3 versions versions. If issues are identified in HTTP/3 then it is recommended that apps temporarily disable it until issues are resolved and the app can be updated.

## HTTP/3 requirements

> HTTP/3 isn't supported everywhere. The requirements are different depending on the operating system hosting Kestrel.
> 
> HTTP/3 isn't enabled by default. Because not all routers, firewalls, and proxies properly support HTTP/3, we recommend configuring HTTP/3 together with HTTP/1.1 and HTTP/2. This can be done by specifying `HttpProtocols.Http1AndHttp2AndHttp3` as an endpoint's supported protocols.
> 
> If the platform that Kestrel is running on doesn't have all the requirements for HTTP/3 then it is automatically disabled.
> 
> For example, `HttpProtocols.Http1AndHttp2AndHttp3` allows Kestrel to enable HTTP/3 on environments where it is supported, with fallbacks for HTTP/1.1 and HTTP/2.

Author: ```suggestion
HTTP/3 isn't enabled by default. Because not all routers, firewalls, and proxies properly support HTTP/3, we recommend configuring HTTP/3 together with HTTP/1.1 and HTTP/2. This can be done by specifying `HttpProtocols.Http1AndHttp2AndHttp3` as an endpoint's supported protocols.

If the platform that Kestrel is running on doesn't have all the requirements for HTTP/3 then it is automatically disabled.

For example, `HttpProtocols.Http1AndHttp2AndHttp3` allows Kestrel to enable HTTP/3 on environments where it is supported, with fallbacks for HTTP/1.1 and HTTP/2.
```

### Windows

* Windows 11 Build 22000 or later.
* TLS 1.3 or later connection.

The preceding Windows 11 Build versions may require the use of a [Windows Insider](https://insider.windows.com) build.

### Linux

On Linux, `libmsquic` is published via Microsoft official Linux package repository `packages.microsoft.com`. In order to consume it, it must be added manually. See [Linux Software Repository for Microsoft Products](/windows-server/administration/linux-package-repository-for-microsoft-software). After adding `libmsquic`, it can be installed via the package manager of your distro, for example, for Ubuntu:

```cmd
apt install libmsquic
```

### macOS

HTTP/3 isn't currently supported on macOS and may be available in a future release.

## Alt-svc

> HTTP/3 is discovered as an upgrade from HTTP/1.1 or HTTP/2 via the `alt-svc` header. That means the first request will normally use HTTP/1.1 or HTTP/2 before switching to HTTP/3.

Author: ```suggestion
HTTP/3 is discovered as an upgrade from HTTP/1.1 or HTTP/2 via the `alt-svc` header. That means the first request will normally use HTTP/1.1 or HTTP/2 before switching to HTTP/3. If HTTP/3 is enabled, Kestrel automatically adds the `alt-svc` header.
```

* Sets `EnableAltSvc` to `true` on Kestrel options.
* Configures port 5001 to use `HttpProtocols.Http1AndHttp2AndHttp3`.

This sample code is specific to .NET 6 Preview 7, and will change in .NET 6 RC 1.

[!code-csharp[](samples/6.x/Http3Sample/Program.cs?name=snippet_UseHttp3&highlight=8)]

For more information on building the host, see the **Set up a host** and **Default builder settings** sections of <xref:fundamentals/host/generic-host#set-up-a-host>.

## Localhost testing

* Browsers do not enable HTTP/3 on localhost/loopback connections: to test with a browser, run the client and server on separate machines.

> * `HttpClient` can be used for localhost/loopback testing in .NET 6 or later. Extra configured is required when using `HttpClient` to make a HTTP/3 request:

Reviewer: ```suggestion
* `HttpClient` can be used for localhost/loopback testing in .NET 6 or later. Extra configured is required when using `HttpClient` to make an HTTP/3 request:
```

  * Set `HttpRequestMessage.Version` to 3.0, or

>   * Set `HttpRequestMessage.VersionPolicy` to `HttpVersionPolicy.RequestVersionOrHigher`.

Reviewer 2: Is this a preview7 specific constraint because of alt-svc? It's working on rc1 without these.
